crashreports/internal/unix/CrashReportContext.cpp

```diff
@@ -0,0 +1,155 @@
+namespace
+{
+
+    //converts parameters map to a string, so that the Crash Reporting program
+    //would understand them
+    std::string StringifyParameters(const std::map<std::string, std::string>& parameters)
+    {
+        std::stringstream stream;
+
+        std::size_t parameterIndex = 0;
+        std::size_t parametersCount = parameters.size();
+        for (auto& pair : parameters)
+        {
+            stream << pair.first.c_str() << "=\"" << pair.second.c_str() << "\"";
+            ++parameterIndex;
+            if (parameterIndex < parametersCount)
+                stream << ",";
+        }
+        return stream.str();
+    }
+
+    //copy contents of src to a raw char dest buffer, 
+    //returns false if dest is not large enough
+    bool StrcpyChecked(char* dest, size_t destsz, const std::string& src)
+    {
+        if(src.length() < destsz)
+        {
+            memcpy(dest, src.c_str(), src.length());
+            dest[src.length()] = '\0';
+            return true;
+        }
+        return false;
+    }
+
+#if defined(__APPLE__)
+
+    static constexpr size_t MaxDumpPathLength{ 4096 };
+    static char DumpPath[MaxDumpPathLength];
+
+    bool DumpCallback(const char* dump_dir, const char* minidump_id, void* context, bool succeeded)
+    {
+        if(succeeded)
+        {
+            const int PathDumpLength = strlen(dump_dir) + strlen("/") + strlen(minidump_id) + strlen(".dmp");
+            if(PathDumpLength < MaxDumpPathLength)
+            {
+                strcpy(DumpPath, dump_dir);
+                strcat(DumpPath, "/");
+                strcat(DumpPath, minidump_id);
+                strcat(DumpPath, ".dmp");
+                auto crashReportContext = static_cast<CrashReportContext*>(context);
+                return SendReport(crashReportContext, DumpPath);
+            }
+            return false;
+        }
+        return succeeded;
+    }
+#else
+    bool DumpCallback(const google_breakpad::MinidumpDescriptor& descriptor, void* context, bool succeeded) 
+    {
+        if(succeeded)
+        {
+            auto crashReportContext = static_cast<CrashReportContext*>(context);
+            return SendReport(crashReportContext, descriptor.path());
+        }
+        return succeeded;
+    }
+#endif
+}
+
+bool CrashReportContext::SetSenderPathUTF8(const std::string& path)
+{
+    return StrcpyChecked(mSenderPath, MaxBufferLength, path + "/" + CRASHREPORTER_PROGRAM_NAME);
+}
+
+bool CrashReportContext::SetReportURL(const std::string& url)
+{
+    return StrcpyChecked(mReportURL, MaxBufferLength, url);
+}
+
+bool CrashReportContext::SetParameters(const std::map<std::string, std::string>& p)
+{
+    auto str = StringifyParameters(p);
+    return StrcpyChecked(mParameters, MaxBufferLength, str);
+}
+
+void CrashReportContext::StartHandler(const std::string& databasePath)
+{
+    //intentinal leak: error hooks may be useful while application is terminating
+    //CrashReportContext data should be alive too...
```

Comment:
Good point.  And that's why it's important to comment that CrashReportContext has a trivial destructor which doesn't change the memory -- not just static allocation of storage.